### ISSUE TITLE
get_aws_account_id() to fetch the role's account id instead of aws profile's account id

### ISSUE THREAD

**Issue by ebarault** [gruntwork-io/terragrunt#379]
I'm using `get_aws_account_id()` to assure uniqueness of the storage bucket per AWS account as in 
```
terragrunt = {
  remote_state {
    backend = "s3"
    config {
      bucket = "mycompany-${get_aws_account_id()}"
    }
  }
}
```

and I run terragrunt commands with the `--terragrunt-iam-role` option to assume a role located in a **member** aws account that belongs to my aws organization.

The profile used to launch the terragrunt commands and assume the role is from the **master** aws account.

I deploy each environment in a dedicated account.

Now, the problem is that `get_aws_account_id()` returns the account id of the **master** account, while the bucket and lock tables are created in the **member** account, meaning that for the next environment I will deploy, the S3 bucket will be the same, as the account id would not change.

Is there a way to make `get_aws_account_id()` return the account id of the assumed role instead?

**Comment by brikis98** [comment]
Ah, interesting find. I guess it is a bit ambiguous which account that's referring to! I guess from a "principle of least surprise" aspect, `--terragrunt-iam-role` feels like it should affect everything globally, so probably that method should assume an IAM role before looking up the account id. I'd welcome a PR to fix that.

**Comment by ebarault** [comment]
I'd be glad to have this landed rather quickly, as i'm starting a new multi-account project and would like to avoid writing down the account ids in the **remote_state** config.

Could you indicate a starting point?

**Comment by brikis98** [comment]
Right here: https://github.com/gruntwork-io/terragrunt/blob/master/config/config_helpers.go#L348

**Comment by ebarault** [comment]
excellent, i'll work on it until the end of week, hopefully.

**Comment by ebarault** [comment]
@brikis98 : could I reach you through gitter (or whatever IM of your choosing) ? I'd just need a quick help with terragrunt's development environment

**Comment by brikis98** [comment]
+1 on https://community.gruntwork.io/. We check it regularly.

**Comment by ebarault** [comment]
ok, just to make sure: I was referring to 'developing terragrunt' (this PR), not 'using terragrunt'

**Comment by brikis98** [comment]
Yup, I hear ya. Same place for that :)

**Comment by ebarault** [comment]
👍

**Comment by ebarault** [comment]
I can't find an existing test for `get_aws_account_id()`, am i missing it ?

**Comment by brikis98** [comment]
It doesn't look like there is one. To some extent, it's hard to create one, as that value would be different for every user running the test.

**Comment by ebarault** [comment]
hmm, just as in the `dynamodb` tests, one would have to configure credentials and the proper role

**Comment by ebarault** [comment]
ok, now i see what could be done with the PR, but before trying a fix : is there a kind of registry inside of terragrunt where we could store / update the sts session created with credentials (local ones or from assumeRoles) instead of creating new session each time a function requiring one kicks in ?

**Comment by brikis98** [comment]
Well, you could probably cache it in the `TerragruntOptions` struct, but I worry that with concurrency (e.g., during `xxx-all` operations), that may lead to problems. Do you have code that calls `get_aws_account_id()` so many times this causes a performance concern? If not, I'd avoid premature optimization.

**Comment by ebarault** [comment]
hmmm, nope i was just asking just in case you'd have one, but we can easily 
- create a new session from local credentials
- get the iam role credentials
- create a new session from iam role credentials
- resolve account from this session

**Comment by brikis98** [comment]
Actually, this method will take care of all of that for you: https://github.com/gruntwork-io/terragrunt/blob/4bcec5557da0ced56a61e53698cb92a72f2fd5a8/aws_helper/config.go#L16. You probably just need to update `getAWSAccountID` to use it.

**Comment by ebarault** [comment]
see https://github.com/gruntwork-io/terragrunt/pull/382

**Comment by brikis98** [comment]
Fixed by #382. Thanks @ebarault!
